Python/compile.c

```diff
@@ -2121,10 +2215,53 @@ compiler_function(struct compiler *c, stmt_ty s, int is_async)
     }
 
     location loc = LOC(s);
+
+    int is_typeparams_in_class = c->u->u_ste->ste_type == ClassBlock;
+    int is_generic = asdl_seq_LEN(typeparams) > 0;
+
+    if (is_generic) {
+        ADDOP(c, loc, PUSH_NULL);
+        // We'll swap in the callable here later.
+        ADDOP_LOAD_CONST(c, loc, Py_None);
+        if (is_typeparams_in_class) {
+            ADDOP(c, loc, LOAD_LOCALS);
+        }
+    }
+
     funcflags = compiler_default_arguments(c, loc, args);
     if (funcflags == -1) {
         return ERROR;
     }
+
+    int num_typeparam_args = 0;
+    _Py_DECLARE_STR(type_params, ".type_params");
+
+    if (is_generic) {
+        PyObject *typeparams_name = PyUnicode_FromFormat("<generic parameters of %U>", name);
+        if (!typeparams_name) {
+            return ERROR;
+        }
+        if (compiler_enter_scope(c, typeparams_name, COMPILER_SCOPE_TYPEPARAMS,
```

Comment:
Good point, the code here leaks the scope in many error paths. Refactored the class, function, and typealias compile functions to avoid this. Also adding a new macro RETURN_IF_ERROR_IN_SCOPE to make this easier.